core/shared/src/main/scala/zio/Ref.scala

```diff
@@ -117,9 +117,32 @@ final class Ref[A] private (private val value: AtomicReference[A]) extends AnyVa
    * implemented in terms of `modify` purely for performance reasons.
    *
    * @param f function to atomically modify the `Ref`
+   * @return `UIO[Unit]`
+   */
+  def update(f: A => A): UIO[Unit] = IO.effectTotal {
+    var loop    = true
+    var next: A = null.asInstanceOf[A]
+
+    while (loop) {
+      val current = value.get
+
+      next = f(current)
+
+      loop = !value.compareAndSet(current, next)
+    }
+
+    ()
+  }
+
+  /**
+   * Atomically modifies the `Ref` with the specified function and returns the
+   * updated value. This is not implemented in terms of `modify` purely for
+   * performance reasons.
+   *
+   * @param f function to atomically modify the `Ref`
    * @return `UIO[A]` modified value of the `Ref`
    */
-  def update(f: A => A): UIO[A] = IO.effectTotal {
+  def updateAndGet(f: A => A): UIO[A] = IO.effectTotal {
```

Comment:
While you're here, do you think it's worth adding `getAndUpdate` too?

Reply:
Yes will do.

Reply:
It could be nice to have `getAndSet` too.